File: src/threat-db/threatStore.ts
import type { Finding, ScanTarget } from "../types.js";
import { InMemoryBackend } from "./backend.js";
import type { ThreatRecord, ThreatStoreBackend } from "./types.js";

const VERSION_KEY = "bundle_version";

/**
 * On-device threat database (PRD §8, FR-8/FR-9).
 *
 * A local cache of threat-intelligence records over a pluggable
 * {@link ThreatStoreBackend} (in-memory or SQLite). Acts as an extra detection
 * input via {@link scan}, returning normalized {@link Finding}s for any stored
 * indicator that matches a scan target.
 */
export class ThreatStore {
  constructor(private readonly backend: ThreatStoreBackend = new InMemoryBackend()) {}

  /** Add or replace threat records. */
  add(records: ThreatRecord[]): void {
    this.backend.upsert(records);
  }

  /** Look up a single indicator, ignoring (and not returning) expired records. */
  lookup(indicator: string, now = Date.now()): ThreatRecord | undefined {
    const rec = this.backend.get(indicator);
    if (!rec) return undefined;
    if (rec.expiresAt !== undefined && rec.expiresAt <= now) return undefined;
    return rec;
  }

  /**
   * Match a scan target's indicators (id, domain, and any program ids in its
   * events) against the store, producing findings for live records.
   */
  scan(target: ScanTarget, now = Date.now()): Finding[] {
    const indicators = new Set<string>();
    indicators.add(target.id);
    if (target.domain) indicators.add(target.domain);
    for (const ev of target.events ?? []) {
      if (ev.programId) indicators.add(ev.programId);
      if (ev.targetAddress) indicators.add(ev.targetAddress);
    }

    const findings: Finding[] = [];
    for (const indicator of indicators) {
      const rec = this.lookup(indicator, now);
      if (!rec) continue;
      findings.push({
        source: "signature",
        ruleId: `THREATDB:${rec.id}`,
        title: `Known threat: ${rec.id}`,
        description: rec.description,
        category: rec.category,
        severity: rec.severity,
        confidence: 0.97, // exact indicator match from curated intel
        evidence: [`${rec.indicatorKind}:${rec.indicator}`, ...(rec.source ? [`source:${rec.source}`] : [])],
      });
    }
    return findings;
  }

  /** Remove expired records; returns how many were purged. */
  purgeExpired(now = Date.now()): number {
    return this.backend.purgeExpired(now);
  }

  /** The applied bundle version (0 if none). */
  get version(): number {
    return Number(this.backend.getMeta(VERSION_KEY) ?? "0");
  }

  set version(v: number) {
    this.backend.setMeta(VERSION_KEY, String(v));
  }

  get size(): number {
    return this.backend.list().length;
  }

  close(): void {
    this.backend.close();
  }
}

File: android/engine/src/main/kotlin/com/daemonblockint/sync/engine/threatdb/ThreatStore.kt
package com.daemonblockint.sync.engine.threatdb

import com.daemonblockint.sync.engine.Finding
import com.daemonblockint.sync.engine.ScanTarget

private const val VERSION_KEY = "bundle_version"

/**
 * On-device threat database (PRD §8, FR-8/FR-9).
 *
 * A local cache of threat-intelligence records over a pluggable
 * ThreatStoreBackend (in-memory or SQLite). Acts as an extra detection
 * input via scan, returning normalized Findings for any stored
 * indicator that matches a scan target.
 */
class ThreatStore(
    private val backend: ThreatStoreBackend = InMemoryBackend(),
) {
    /** Add or replace threat records. */
    fun add(records: List<ThreatRecord>) {
        backend.upsert(records)
    }

    /** Look up a single indicator, ignoring (and not returning) expired records. */
    fun lookup(indicator: String, now: Long = System.currentTimeMillis()): ThreatRecord? {
        val rec = backend.get(indicator) ?: return null
        if (rec.expiresAt != null && rec.expiresAt <= now) return null
        return rec
    }

    /**
     * Match a scan target's indicators (id, domain, and any program ids in its
     * events) against the store, producing findings for live records.
     */
    fun scan(target: ScanTarget, now: Long = System.currentTimeMillis()): List<Finding> {
        val indicators = linkedSetOf<String>()
        indicators.add(target.id)
        target.domain?.let { indicators.add(it) }
        for (ev in target.events) {
            ev.programId?.let { indicators.add(it) }
            ev.targetAddress?.let { indicators.add(it) }
        }

        val findings = mutableListOf<Finding>()
        for (indicator in indicators) {
            val rec = lookup(indicator, now) ?: continue
            findings.add(
                Finding(
                    source = Finding.Source.SIGNATURE,
                    ruleId = "THREATDB:${rec.id}",
                    title = "Known threat: ${rec.id}",
                    description = rec.description,
                    category = rec.category,
                    severity = rec.severity,
                    confidence = 0.97,
                    evidence = buildList {
                        add("${rec.indicatorKind}:${rec.indicator}")
                        rec.source?.let { add("source:$it") }
                    },
                ),
            )
        }
        return findings
    }

    /** Remove expired records; returns how many were purged. */
    fun purgeExpired(now: Long = System.currentTimeMillis()): Int =
        backend.purgeExpired(now)

    /** The applied bundle version (0 if none). */
    var version: Int
        get() = backend.getMeta(VERSION_KEY)?.toIntOrNull() ?: 0
        set(v) { backend.setMeta(VERSION_KEY, v.toString()) }

    val size: Int get() = backend.list().size

    fun close() { backend.close() }
}
